source/MaterialXRenderGlsl/TextureBaker.cpp

```diff
@@ -101,20 +103,53 @@ TextureBaker::TextureBaker(unsigned int width, unsigned int height, Image::BaseT
     _frameCaptureImage->createResourceBuffer();
 }
 
-FilePath TextureBaker::generateTextureFilename(OutputPtr output, const string& shaderName, const string& udim)
+size_t TextureBaker::findVarInTemplate(const string& filename, const string& var, size_t start)
 {
-    string outputName = createValidName(output->getNamePath());
-    string shaderSuffix = shaderName.empty() ? EMPTY_STRING : "_" + shaderName;
-    string udimSuffix = udim.empty() ? EMPTY_STRING : "_" + udim;
-    std::string bakedImageName; 
-    bakedImageName = outputName + shaderSuffix + BAKED_POSTFIX + udimSuffix;
-    if (_hashImageNames)
-    {
-        std::stringstream hashStream;
-        hashStream << std::hash<std::string>{}(bakedImageName);
-        bakedImageName = hashStream.str();
-    }
-    return FilePath(bakedImageName + "." + _extension);
+    size_t i = filename.find(var, start);
+    if (var == "$UDIM" && i != string::npos)
+    {
+        size_t udimPrefix = filename.find("$UDIMPREFIX", start);
+        if (i == udimPrefix)
+        {
+            i = filename.find(var, i + 1);
+        }
+    }
+    return i;
+}
+
+FilePath TextureBaker::generateTextureFilename(const StringMap& filenameTemplateMap)
+{
+    string bakedImageName = _textureFilenameTemplate;
+
+    for (auto& pair : filenameTemplateMap)
+    {
+        string replacement = (_texTemplateOverrides.count(pair.first)) ?
+            _texTemplateOverrides[pair.first] : pair.second;
+        replacement = (filenameTemplateMap.at("$UDIM").empty() && pair.first == "$UDIMPREFIX") ?
+            EMPTY_STRING : replacement;
+
+        for (size_t i = 0; (i = findVarInTemplate(bakedImageName, pair.first, i)) != string::npos; i++)
+        {
+            bakedImageName.replace(i, pair.first.length(), replacement);
+        }
+    }
+    return _outputImagePath / bakedImageName;
+}
+
+StringMap TextureBaker::initializeFileTemplateMap(InputPtr input, NodePtr shader, const string& udim)
+{
+    string inputName = input->getName();
+    FilePath assetPath = FilePath(shader->getActiveSourceUri());
+    assetPath.removeExtension();
+    StringMap filenameTemplateMap;
+    filenameTemplateMap["$ASSET"] = assetPath.getBaseName();
+    filenameTemplateMap["$INPUT"] = _bakedInputOverlapMap.count(inputName)? _bakedInputOverlapMap[inputName] : inputName;
```

Comment:
To simplify this new logic, you might consider naming the new map `_bakedInputMap`, and setting it consistently for all baked inputs, not just those that share their baked texture with another input.  That would allow this logic to be written more simply as:

```
filenameTemplateMap["$INPUT"] = _bakedInputMap.count(inputName);
```